File: app/audit.py
"""Agent Auditor — applique le referentiel fiscal et chiffre l'exposition.

Aucun appel au modele. Chaque anomalie porte :
  - la regle du referentiel qui la fonde (regles-fiscales.md)
  - une exposition en dirhams calculee en Decimal
  - un niveau de confiance, abaisse par les rejets deja prononces (EX-06)

Convention d'exposition : le risque, c'est la TVA que l'administration peut
refuser en deduction, ou le redressement d'assiette. Jamais le TTC entier, sauf
doublon paye deux fois.
"""
from datetime import date, timedelta

from app.money import q, d, tva_attendue
from app.referentiel import PERIODE_DEBUT, PERIODE_FIN
from app.reconcile import lire_date

ECART_DOUBLON = timedelta(days=7)
FACTEUR_ABERRANT = d("10")

FAMILLES = {
    "doublon_exact": ("Doublon exact", "regle 6"),
    "doublon_probable": ("Doublon probable", "regle 5"),
    "tva_erronee": ("TVA au mauvais taux", "taux par categorie"),
    "tva_incoherente": ("TVA incoherente avec la base HT", "regle 2"),
    "hors_periode": ("Piece hors exercice", "regle 4"),
    "tiers_inconnu": ("Tiers absent du referentiel", "regle 3"),
    "montant_aberrant": ("Montant aberrant", "regle 7"),
    "ice_manquant": ("ICE fournisseur absent", "regle 3"),
    "document_non_traite": ("Document non exploitable", "EX-08"),
}


def _anomalie(piece, famille, exposition, confiance, detail, piece_liee=None):
    libelle, regle = FAMILLES[famille]
    return {
        "doc_id": piece["doc_id"],
        "fichier": piece.get("fichier"),
        "famille": famille,
        "libelle": libelle,
        "regle": regle,
        "fournisseur": piece.get("tiers_libelle") or piece.get("fournisseur"),
        "date": piece.get("date"),
        "numero": piece.get("numero"),
        "ttc": piece.get("ttc"),
        "exposition_mad": q(exposition),
        "confiance": round(confiance, 2),
        "detail": detail,
        "piece_liee": piece_liee,
        "statut_revue": "a_arbitrer",
    }


# ------------------------------------------------------------ controles piece

def controler_piece(piece: dict) -> list:
    anomalies = []
    fiche = piece.get("fiche_fournisseur")
    ht, tva, ttc = piece.get("ht"), piece.get("tva"), piece.get("ttc")
    taux = piece.get("taux_tva")
    tva_abs = abs(tva) if tva is not None else None

    if piece.get("statut") != "traite":
        anomalies.append(_anomalie(
            piece, "document_non_traite", d("0"), 1.0,
            f"Lecture impossible : {piece.get('motif')}. Aucun montant n'a ete "
            f"retenu, la piece reste dans la file humaine."))
        return anomalies

    # tiers absent du referentiel (regle 3)
    if not fiche:
        anomalies.append(_anomalie(
            piece, "tiers_inconnu", tva_abs or d("0"), 0.9,
            f"'{piece.get('tiers_libelle')}' ne figure pas dans le referentiel "
            f"fournisseurs. La deduction de TVA est a justifier."))

    # ICE fournisseur (regle 1 et 3)
    if not piece.get("ice_fournisseur"):
        anomalies.append(_anomalie(
            piece, "ice_manquant", tva_abs or d("0"), 0.95,
            "Aucun ICE fournisseur lisible sur la piece : elle n'ouvre pas "
            "droit a deduction en l'etat."))

    # taux de TVA vs categorie du fournisseur
    if fiche and taux is not None:
        habituel = fiche["taux_tva_habituel"]
        if taux != habituel and ht is not None:
            ecart = abs(tva_attendue(abs(ht), habituel) - (tva_abs or d("0")))
            anomalies.append(_anomalie(
                piece, "tva_erronee", ecart, 0.95,
                f"TVA a {taux}% alors que la categorie '{fiche['categorie']}' "
                f"releve du taux de {habituel}%. Redressement d'assiette calcule "
                f"sur une base HT de {q(abs(ht))} MAD."))

    # coherence interne TVA = HT x taux
    if taux is not None and ht is not None and tva is not None:
        ecart = abs(abs(tva) - tva_attendue(abs(ht), taux))
        if ecart > d("0.02"):
            anomalies.append(_anomalie(
                piece, "tva_incoherente", ecart, 0.85,
                f"La TVA portee ({q(abs(tva))} MAD) ne correspond pas a "
                f"{taux}% de la base HT ({tva_attendue(abs(ht), taux)} MAD)."))

    # periode (regle 4)
    if piece.get("date") and not (PERIODE_DEBUT <= piece["date"] <= PERIODE_FIN):
        anomalies.append(_anomalie(
            piece, "hors_periode", tva_abs or d("0"), 1.0,
            f"Piece datee du {piece['date']}, hors exercice "
            f"{PERIODE_DEBUT} au {PERIODE_FIN} : non imputable."))

    # montant aberrant (regle 7)
    if fiche and ttc is not None:
        plafond = q(fiche["montant_moyen_ttc_mad"] * FACTEUR_ABERRANT)
        if abs(ttc) > plafond:
            anomalies.append(_anomalie(
                piece, "montant_aberrant", tva_abs or d("0"), 0.8,
                f"TTC de {q(abs(ttc))} MAD contre une moyenne historique de "
                f"{fiche['montant_moyen_ttc_mad']} MAD (plafond {plafond} MAD)."))
    return anomalies


# --------------------------------------------------------------- doublons

def detecter_doublons(pieces: list) -> list:
    """Regles 5 et 6. On ne fusionne jamais : on signale avec une confiance."""
    anomalies = []
    retenues = [p for p in pieces if p.get("statut") == "traite"
                and p.get("ttc") is not None and p.get("type_piece") == "facture"]
    par_tiers = {}
    for p in retenues:
        par_tiers.setdefault(p.get("tiers_libelle"), []).append(p)

    for tiers, groupe in par_tiers.items():
        groupe = sorted(groupe, key=lambda p: (p.get("date") or "", p["doc_id"]))
        for i, a in enumerate(groupe):
            for b in groupe[i + 1:]:
                if q(a["ttc"]) != q(b["ttc"]):
                    continue
                da, db = lire_date(a.get("date")), lire_date(b.get("date"))
                if not da or not db:
                    continue
                exact = (a.get("numero") == b.get("numero") and da == db)
                if exact:
                    anomalies.append(_anomalie(
                        b, "doublon_exact", abs(b.get("tva") or d("0")), 1.0,
                        f"Piece identique a {a['doc_id']} : meme numero "
                        f"{a.get('numero')}, meme date, meme TTC "
                        f"{q(a['ttc'])} MAD. TVA deduite deux fois.",
                        piece_liee=a["doc_id"]))
                elif abs((db - da).days) < ECART_DOUBLON.days:
                    anomalies.append(_anomalie(
                        b, "doublon_probable", abs(b.get("tva") or d("0")), 0.75,
                        f"Meme tiers et meme TTC ({q(b['ttc'])} MAD) que "
                        f"{a['doc_id']}, a {abs((db - da).days)} jours d'ecart "
                        f"(numeros {a.get('numero')} et {b.get('numero')}). "
                        f"A confirmer avant rejet : aucune fusion automatique.",
                        piece_liee=a["doc_id"]))
    return anomalies


# ------------------------------------------------- boucle de revue humaine

def appliquer_decisions(anomalies: list, decisions: list) -> list:
    """EX-06 : un rejet deja prononce sur un couple (tiers, famille) abaisse la
    confiance des cas suivants ; trois rejets les masquent par defaut."""
    compte = {}
    for dec in decisions:
        cle = (dec["fournisseur"], dec["famille"])
        compte.setdefault(cle, {"valide": 0, "rejete": 0})
        compte[cle][dec["verdict"]] = compte[cle].get(dec["verdict"], 0) + 1

    for a in anomalies:
        cle = (a["fournisseur"], a["famille"])
        stats = compte.get(cle)
        if not stats:
            continue
        rejets, validations = stats.get("rejete", 0), stats.get("valide", 0)
        if rejets:
            a["confiance"] = round(max(0.05, a["confiance"] - 0.25 * rejets), 2)
            a["apprentissage"] = (f"{rejets} rejet(s) deja prononce(s) sur ce "
                                  f"motif pour ce tiers")
            if rejets >= 3:
                a["statut_revue"] = "masquee_apres_rejets"
        if validations:
            a["confiance"] = round(min(1.0, a["confiance"] + 0.05 * validations), 2)
    return anomalies


def exposition_totale(anomalies: list):
    """Une piece peut porter plusieurs anomalies — ICE absent et montant
    aberrant, par exemple — mais c'est la meme TVA qui est en jeu. Le total
    retient donc, par piece, la plus forte exposition et non leur somme."""
    par_piece = {}
    for a in anomalies:
        montant = a["exposition_mad"] or d("0")
        if montant > par_piece.get(a["doc_id"], d("0")):
            par_piece[a["doc_id"]] = montant
    return q(sum(par_piece.values(), d("0")))


def auditer(pieces: list, decisions: list = None) -> list:
    anomalies = []
    for piece in pieces:
        anomalies.extend(controler_piece(piece))
    anomalies.extend(detecter_doublons(pieces))
    anomalies = appliquer_decisions(anomalies, decisions or [])
    anomalies.sort(key=lambda a: (-(a["exposition_mad"] or d("0")), -a["confiance"]))
    return anomalies

File: api/src/audit.ts
/**
 * Agent Auditor — applique le referentiel fiscal et chiffre l'exposition.
 *
 * Aucun appel au modele. Chaque anomalie porte la regle qui la fonde, une
 * exposition en dirhams calculee en Decimal, et un niveau de confiance abaisse
 * par les rejets deja prononces (EX-06).
 *
 * Convention d'exposition : le risque, c'est la TVA que l'administration peut
 * refuser en deduction, ou le redressement d'assiette. Jamais le TTC entier.
 */
import { Decimal } from "decimal.js";
import { q, tvaAttendue, ZERO, type Montant } from "./money.js";
import type { Piece } from "./ingest.js";
import { lireDate } from "./reconcile.js";
import { PERIODE_DEBUT, PERIODE_FIN } from "./referentiel.js";

const ECART_DOUBLON_JOURS = 7;
const FACTEUR_ABERRANT = new Decimal(10);

export type Famille =
  | "doublon_exact"
  | "doublon_probable"
  | "tva_erronee"
  | "tva_incoherente"
  | "hors_periode"
  | "tiers_inconnu"
  | "montant_aberrant"
  | "ice_manquant"
  | "document_non_traite";

const FAMILLES: Record<Famille, { libelle: string; regle: string }> = {
  doublon_exact: { libelle: "Doublon exact", regle: "regle 6" },
  doublon_probable: { libelle: "Doublon probable", regle: "regle 5" },
  tva_erronee: { libelle: "TVA au mauvais taux", regle: "taux par categorie" },
  tva_incoherente: { libelle: "TVA incoherente avec la base HT", regle: "regle 2" },
  hors_periode: { libelle: "Piece hors exercice", regle: "regle 4" },
  tiers_inconnu: { libelle: "Tiers absent du referentiel", regle: "regle 3" },
  montant_aberrant: { libelle: "Montant aberrant", regle: "regle 7" },
  ice_manquant: { libelle: "ICE fournisseur absent", regle: "regle 3" },
  document_non_traite: { libelle: "Document non exploitable", regle: "EX-08" },
};

export interface Anomalie {
  docId: string;
  fichier: string;
  famille: Famille;
  libelle: string;
  regle: string;
  fournisseur: string | null;
  date: string | null;
  numero: string | null;
  ttc: string | null;
  expositionMad: Montant;
  confiance: number;
  detail: string;
  pieceLiee: string | null;
  statutRevue: "a_arbitrer" | "masquee_apres_rejets";
  apprentissage?: string;
  action?: string;
  actionSource?: string;
}

export interface Decision {
  fournisseur: string | null;
  famille: string;
  verdict: "valide" | "rejete";
}

function anomalie(
  piece: Piece,
  famille: Famille,
  exposition: Montant,
  confiance: number,
  detail: string,
  pieceLiee: string | null = null
): Anomalie {
  return {
    docId: piece.docId,
    fichier: piece.fichier,
    famille,
    libelle: FAMILLES[famille].libelle,
    regle: FAMILLES[famille].regle,
    fournisseur: piece.tiersLibelle,
    date: piece.date,
    numero: piece.numero,
    ttc: piece.ttc ? piece.ttc.toFixed(2) : null,
    expositionMad: q(exposition),
    confiance: Math.round(confiance * 100) / 100,
    detail,
    pieceLiee,
    statutRevue: "a_arbitrer",
  };
}

// ------------------------------------------------------------ controles piece

export function controlerPiece(piece: Piece): Anomalie[] {
  const anomalies: Anomalie[] = [];

  if (piece.statut !== "traite") {
    anomalies.push(
      anomalie(
        piece,
        "document_non_traite",
        ZERO,
        1,
        `Lecture impossible : ${piece.motif}. Aucun montant n'a ete retenu, ` +
          `la piece reste dans la file humaine.`
      )
    );
    return anomalies;
  }

  const { fiche, ht, tva, ttc, tauxTva } = piece;
  const tvaAbs = tva ? tva.abs() : null;

  // Tiers absent du referentiel (regle 3).
  if (!fiche) {
    anomalies.push(
      anomalie(
        piece,
        "tiers_inconnu",
        tvaAbs ?? ZERO,
        0.9,
        `« ${piece.tiersLibelle} » ne figure pas dans le referentiel ` +
          `fournisseurs. La deduction de TVA est a justifier.`
      )
    );
  }

  // ICE fournisseur (regles 1 et 3).
  if (!piece.iceFournisseur) {
    anomalies.push(
      anomalie(
        piece,
        "ice_manquant",
        tvaAbs ?? ZERO,
        0.95,
        "Aucun ICE fournisseur lisible sur la piece : elle n'ouvre pas droit " +
          "a deduction en l'etat."
      )
    );
  }

  // Taux de TVA face a la categorie du fournisseur.
  if (fiche && tauxTva !== null && ht) {
    if (tauxTva !== fiche.tauxTvaHabituel) {
      const ecart = tvaAttendue(ht.abs(), fiche.tauxTvaHabituel)
        .minus(tvaAbs ?? ZERO)
        .abs();
      anomalies.push(
        anomalie(
          piece,
          "tva_erronee",
          ecart,
          0.95,
          `TVA a ${tauxTva} % alors que la categorie « ${fiche.categorie} » ` +
            `releve du taux de ${fiche.tauxTvaHabituel} %. Redressement ` +
            `d'assiette calcule sur une base HT de ${q(ht.abs()).toFixed(2)} MAD.`
        )
      );
    }
  }

  // Coherence interne : TVA = HT x taux.
  if (tauxTva !== null && ht && tva) {
    const ecart = tva.abs().minus(tvaAttendue(ht.abs(), tauxTva)).abs();
    if (ecart.greaterThan("0.02")) {
      anomalies.push(
        anomalie(
          piece,
          "tva_incoherente",
          ecart,
          0.85,
          `La TVA portee (${q(tva.abs()).toFixed(2)} MAD) ne correspond pas a ` +
            `${tauxTva} % de la base HT (${tvaAttendue(ht.abs(), tauxTva).toFixed(2)} MAD).`
        )
      );
    }
  }

  // Periode (regle 4).
  if (piece.date && (piece.date < PERIODE_DEBUT || piece.date > PERIODE_FIN)) {
    anomalies.push(
      anomalie(
        piece,
        "hors_periode",
        tvaAbs ?? ZERO,
        1,
        `Piece datee du ${piece.date}, hors exercice ${PERIODE_DEBUT} au ` +
          `${PERIODE_FIN} : non imputable.`
      )
    );
  }

  // Montant aberrant (regle 7).
  if (fiche && ttc) {
    const plafond = q(fiche.montantMoyenTtc.times(FACTEUR_ABERRANT));
    if (ttc.abs().greaterThan(plafond)) {
      anomalies.push(
        anomalie(
          piece,
          "montant_aberrant",
          tvaAbs ?? ZERO,
          0.8,
          `TTC de ${q(ttc.abs()).toFixed(2)} MAD contre une moyenne historique ` +
            `de ${fiche.montantMoyenTtc.toFixed(2)} MAD (plafond ${plafond.toFixed(2)} MAD).`
        )
      );
    }
  }
  return anomalies;
}

// ------------------------------------------------------------------ doublons

/** Regles 5 et 6. On ne fusionne jamais : on signale avec une confiance. */
export function detecterDoublons(pieces: Piece[]): Anomalie[] {
  const anomalies: Anomalie[] = [];
  const parTiers = new Map<string, Piece[]>();

  for (const piece of pieces) {
    if (piece.statut !== "traite" || !piece.ttc || piece.typePiece !== "facture") continue;
    const cle = piece.tiersLibelle ?? "";
    parTiers.set(cle, [...(parTiers.get(cle) ?? []), piece]);
  }

  for (const groupe of parTiers.values()) {
    const ordonne = [...groupe].sort((a, b) =>
      `${a.date}${a.docId}`.localeCompare(`${b.date}${b.docId}`)
    );
    for (let i = 0; i < ordonne.length; i += 1) {
      for (let j = i + 1; j < ordonne.length; j += 1) {
        const a = ordonne[i];
        const b = ordonne[j];
        if (!q(a.ttc!).equals(q(b.ttc!))) continue;
        const da = lireDate(a.date);
        const db = lireDate(b.date);
        if (!da || !db) continue;
        const ecart = Math.abs(Math.round((db.getTime() - da.getTime()) / 86_400_000));

        if (a.numero === b.numero && ecart === 0) {
          anomalies.push(
            anomalie(
              b,
              "doublon_exact",
              b.tva?.abs() ?? ZERO,
              1,
              `Piece identique a ${a.docId} : meme numero ${a.numero}, meme ` +
                `date, meme TTC ${q(a.ttc!).toFixed(2)} MAD. TVA deduite deux fois.`,
              a.docId
            )
          );
        } else if (ecart < ECART_DOUBLON_JOURS) {
          anomalies.push(
            anomalie(
              b,
              "doublon_probable",
              b.tva?.abs() ?? ZERO,
              0.75,
              `Meme tiers et meme TTC (${q(b.ttc!).toFixed(2)} MAD) que ` +
                `${a.docId}, a ${ecart} jours d'ecart (numeros ${a.numero} et ` +
                `${b.numero}). A confirmer avant rejet : aucune fusion automatique.`,
              a.docId
            )
          );
        }
      }
    }
  }
  return anomalies;
}

// ------------------------------------------------- boucle de revue humaine

/**
 * EX-06 : un rejet deja prononce sur un couple (tiers, famille) abaisse la
 * confiance des cas suivants ; trois rejets les masquent par defaut.
 */
export function appliquerDecisions(anomalies: Anomalie[], decisions: Decision[]): Anomalie[] {
  const compte = new Map<string, { valide: number; rejete: number }>();
  for (const d of decisions) {
    const cle = `${d.fournisseur}::${d.famille}`;
    const stats = compte.get(cle) ?? { valide: 0, rejete: 0 };
    stats[d.verdict] += 1;
    compte.set(cle, stats);
  }

  for (const a of anomalies) {
    const stats = compte.get(`${a.fournisseur}::${a.famille}`);
    if (!stats) continue;
    if (stats.rejete > 0) {
      a.confiance = Math.round(Math.max(0.05, a.confiance - 0.25 * stats.rejete) * 100) / 100;
      a.apprentissage = `${stats.rejete} rejet(s) deja prononce(s) sur ce motif pour ce tiers`;
      if (stats.rejete >= 3) a.statutRevue = "masquee_apres_rejets";
    }
    if (stats.valide > 0) {
      a.confiance = Math.round(Math.min(1, a.confiance + 0.05 * stats.valide) * 100) / 100;
    }
  }
  return anomalies;
}

/**
 * Une piece peut porter plusieurs anomalies — ICE absent et montant aberrant,
 * par exemple — mais c'est la meme TVA qui est en jeu. Le total retient donc,
 * par piece, la plus forte exposition et non leur somme.
 */
export function expositionTotale(anomalies: Anomalie[]): Montant {
  const parPiece = new Map<string, Montant>();
  for (const a of anomalies) {
    const actuelle = parPiece.get(a.docId);
    if (!actuelle || a.expositionMad.greaterThan(actuelle)) {
      parPiece.set(a.docId, a.expositionMad);
    }
  }
  return q([...parPiece.values()].reduce((total, m) => total.plus(m), ZERO));
}

export function auditer(pieces: Piece[], decisions: Decision[] = []): Anomalie[] {
  const anomalies = [
    ...pieces.flatMap(controlerPiece),
    ...detecterDoublons(pieces),
  ];
  appliquerDecisions(anomalies, decisions);
  anomalies.sort((x, y) => {
    const ecart = y.expositionMad.comparedTo(x.expositionMad);
    return ecart !== 0 ? ecart : y.confiance - x.confiance;
  });
  return anomalies;
}
